### PR TITLE
Send error to Conttroller on container failed to create

### PR DESCRIPTION
**Pull request by NehaNaithani**

### CODE REVIEW

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java [hunk 1]
```diff
@@ -185,8 +187,22 @@ public void startContainer(Microservice microservice) throws NotFoundException,
 //
 //		if (totalMemory - jvmMemory < requiredMemory)
 //			throw new Exception("Not enough memory to start the container");
-    	LoggingService.logDebug(MODULE_NAME , "start Container");
-        dockerClient.startContainerCmd(microservice.getContainerId()).exec();
+        try {
+            LoggingService.logDebug(MODULE_NAME , "start Container");
```
Comment (sergerad) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java:191]: Why catch all 3 exceptions when the logic for each is the same?

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java [hunk 2]
```diff
@@ -311,21 +327,33 @@ private long getStartedTime(String startedTime) {
...
+            ContainerState containerState = inspectInfo.getState();
+            if (containerState != null) {
+                if (containerState.getStartedAt() != null) {
+                    result.setStartTime(getStartedTime(containerState.getStartedAt()));
+                }
 
-            result.setContainerId(containerId);
-            result.setUsage(containerId);
+                MicroserviceState microserviceState = containerToMicroserviceState(containerState);
+                result.setStatus(MicroserviceState.EXITING.equals(microserviceState) && RestartStuckChecker.isStuck(microServiceUuid)
```
Comment (sergerad) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java]: Rewrite this ternary logic! Make another function that maps the boolean inputs to and output if it makes it tidier

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java [hunk 3]
```diff
@@ -311,21 +327,33 @@ private long getStartedTime(String startedTime) {
...
+                        ? MicroserviceState.STUCK_IN_RESTART
+                        : microserviceState);
+
+                result.setContainerId(containerId);
+                result.setUsage(containerId);
+                MicroserviceStatus existingStatus = StatusReporter.setProcessManagerStatus().getMicroserviceStatus(microServiceUuid);
+                result.setPercentage(existingStatus.getPercentage());
+                result.setErrorMessage(existingStatus.getErrorMessage());
+            }
+        } catch (NotFoundException e) {
```
Comment (sergerad) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java]: Again why catch 2 exceptions when logic is the same?

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java [hunk 4]
```diff
@@ -696,8 +727,23 @@ public String createContainer(Microservice microservice, String host) throws Not
             cmd = cmd.withCmd(microservice.getArgs());
         }
         cmd = cmd.withHostConfig(hostConfig);
-        CreateContainerResponse resp = cmd.exec();
-        LoggingService.logInfo(MODULE_NAME ,String.format("Container created \"%s\" ", microservice.getImageName()));
+        CreateContainerResponse resp;
+        try {
+            resp = cmd.exec();
```
Comment (sergerad) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java:730]: again multiple catches for same logic

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java [hunk 5]
```diff
@@ -184,8 +185,11 @@ private void handleLatestMicroservices() {
 				Optional<Container> containerOptional = docker.getContainer(microservice.getMicroserviceUuid());
 
 				if (!containerOptional.isPresent() && !microservice.isDelete()) {
-					StatusReporter.setProcessManagerStatus().setMicroservicesState(microservice.getMicroserviceUuid(), MicroserviceState.QUEUED);
-					addMicroservice(microservice);
+					if (StatusReporter.getProcessManagerStatus().getMicroserviceStatus(microservice.getMicroserviceUuid()) != null
```
Comment (sergerad) [inline iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/ProcessManager.java]: These lines of code are too long, shorten by assigning to variables before final evaluation

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java (hunk 1 above)
Comment (NehaNaithani) [reply iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java:191]: I think we should always catch the checked exception and the exception in the end. Although I agree the logic is same, will update the error messages.

Comment (sergerad) [reply iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java:191]: If you are just logging the exception message, you only need to catch the most general exception. You don't have any exception-specific behaviour

File: iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java (hunk 3 above)
Comment (NehaNaithani) [reply iofog-agent-daemon/src/main/java/org/eclipse/iofog/process_manager/DockerUtil.java]: updated

**Review by sergerad — APPROVED** [review]

**Review by baghbidi — APPROVED** [review]

**Pull request merged**